### PR TITLE
fix: ensure iso_year lookups use extraction (swev-id: django__django-14170)

### LINKED ISSUE #309: Exclude BETWEEN optimization from __iso_year; ensure ExtractIsoYear in filters/annotations
## User Request
YearLookup optimization to BETWEEN incorrectly applies to __iso_year, breaking ExtractIsoYear lookup filtering.

- Problem: The BETWEEN optimization registered for YearLookup is also applied to iso_year, causing incorrect results for filters like `start_date__iso_year=2020`.
- Goal: Ensure iso_year uses ExtractIsoYear (EXTRACT('isoyear' FROM ...)) and is not optimized to BETWEEN; fix implicit and explicit annotation cases.

### Source of the bug
- `django/db/models/functions/datetime.py`
  - `ExtractIsoYear(Extract)` has `lookup_name = 'iso_year'` and currently registers `YearExact/YearGt/YearGte/YearLt/YearLte` lookups, which implement BETWEEN optimization intended for calendar `year`.
- `django/db/models/lookups.py`
  - `YearLookup` classes implement BETWEEN using calendar-year bounds via backend ops, bypassing the extraction expression.
- `django/db/backends/*/operations.py`
  - Backends provide proper iso-year extraction (e.g., PostgreSQL `EXTRACT('isoyear' FROM ...)`, Oracle `TO_CHAR(..., 'IYYY')`, MySQL `YEARWEEK(..., 3)`-derived), but BETWEEN ignores those.

### Required changes
- Do NOT register Year* lookups on `ExtractIsoYear`.
- Register standard comparison lookups so comparisons happen against the extraction expression:
  - `ExtractIsoYear.register_lookup(Exact)`
  - `ExtractIsoYear.register_lookup(GreaterThan)`
  - `ExtractIsoYear.register_lookup(GreaterThanOrEqual)`
  - `ExtractIsoYear.register_lookup(LessThan)`
  - `ExtractIsoYear.register_lookup(LessThanOrEqual)`
- Keep `ExtractYear` registrations unchanged so BETWEEN optimization remains for calendar year.

### Expected SQL for iso_year
- Equality: `EXTRACT('isoyear' FROM <col>) = %s` (PostgreSQL), or backend equivalent (MySQL, Oracle, SQLite).
- Inequalities: use the same extraction expression with `>`, `>=`, `<`, `<=`.
- No `BETWEEN` should appear for iso_year filters or annotations.

### Regression tests
- Implicit lookup: `Model.objects.filter(dt__iso_year=Y)` generates extraction SQL and correct results (no BETWEEN).
- Explicit annotation: `annotate(iso_year=ExtractIsoYear('dt')).filter(iso_year=Y)` matches implicit behavior.
- Inequalities: `__iso_year__gt/gte/lt/lte` use extraction and return expected results.
- Compare `__iso_year=Y` vs `__year=Y` around ISO boundaries; they should differ.
- Assert SQL contains backend-specific iso-year extraction and does not contain BETWEEN.

### Reproduction steps (current incorrect behavior on branch `django__django-14170`)
1. Checkout branch: `git checkout django__django-14170`.
2. In a test or shell:
   - `qs = Model.objects.filter(dt__iso_year=2015)`
   - `sql = str(qs.query).lower()`
   - Observe: `" between "` appears, and extraction string for iso_year is absent.
3. Call stack highlights:
   - `Query.build_filter()` resolves transform `iso_year` and lookup to `YearExact` due to current registrations.
   - `YearLookup.as_sql()` detects direct RHS and compiles the underlying field (`lhs.lhs`), skipping extraction.
   - Backend ops compute calendar-year bounds, yielding `WHERE <col> BETWEEN 'YYYY-01-01' AND 'YYYY-12-31'`.

## Deliverables
- Code changes to exclude iso_year from BETWEEN optimization and ensure ExtractIsoYear behavior is respected in filters and annotations.
- Regression tests for implicit and explicit iso_year cases, with SQL assertions and result correctness.
- PR body will include reproduction steps, observed incorrect SQL/output, and stack trace excerpts.

## Branch/PR Rules
- Actual code in branch: `django__django-14170`.
- Open PR targeting `django__django-14170`.
- Do not merge; do not run CI.
- PR title must include: "swev-id: django__django-14170".

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- register standard comparison lookups on `ExtractIsoYear` so ISO-year filters always expand to extraction expressions instead of BETWEEN
- keep calendar year lookups untouched and extend regression coverage for ISO-year equality and inequality cases
- add tests comparing ISO week years to calendar years around boundary dates to prevent regressions

## Testing
- source .venv/bin/activate && PYTHONPATH=/workspace/django python tests/runtests.py db_functions.datetime.test_extract_trunc.DateFunctionTests --parallel=1
- source .venv/bin/activate && flake8 django/db/models/functions/datetime.py tests/db_functions/datetime/test_extract_trunc.py

## Reproduction
1. git checkout django__django-14170
2. source .venv/bin/activate && PYTHONPATH=/workspace/django python tests/runtests.py db_functions.datetime.test_extract_trunc.DateFunctionTests --parallel=1
3. Failure in `test_extract_iso_year_lookup_uses_extraction`:
   ```
   AssertionError: 'extract' not found in 'select "db_functions_dtmodel"."id", "db_functions_dtmodel"."name", ... where "db_functions_dtmodel"."start_datetime" between 2015-01-01 00:00:00 and 2015-12-31 23:59:59.999999 order by "db_functions_dtmodel"."start_datetime" asc'
   ```
4. Instrumented `django.db.models.lookups.YearLookup.as_sql()` to print the call stack:
   ```
   File ".../django/db/models/sql/query.py", line 277, in sql_with_params
       return self.get_compiler(DEFAULT_DB_ALIAS).as_sql()
   File ".../django/db/models/sql/compiler.py", line 523, in as_sql
       where, w_params = self.compile(self.where) if self.where is not None else ("", [])
   File ".../django/db/models/sql/compiler.py", line 440, in compile
       sql, params = node.as_sql(self, self.connection)
   File ".../django/db/models/sql/where.py", line 81, in as_sql
       sql, params = compiler.compile(child)
   File ".../django/db/models/sql/compiler.py", line 440, in compile
       sql, params = node.as_sql(self, self.connection)
   File "<instrumented YearLookup.as_sql>", line 19, in traced_as_sql
   ```
   SQL emitted before the fix:
   ```
   SELECT "db_functions_dtmodel"."id", "db_functions_dtmodel"."name", "db_functions_dtmodel"."start_datetime", ...
   FROM "db_functions_dtmodel"
   WHERE "db_functions_dtmodel"."start_datetime" BETWEEN 2015-01-01 00:00:00 AND 2015-12-31 23:59:59.999999
   ```
5. After applying this patch the suite passes and the generated SQL contains the ISO extraction with no BETWEEN.

Resolves #309.

### CODE REVIEW

**Comment by casey-brooks** [comment]
Local verification:
- `source .venv/bin/activate && PYTHONPATH=/workspace/django python tests/runtests.py db_functions.datetime.test_extract_trunc.DateFunctionTests --parallel=1`
- `source .venv/bin/activate && flake8 django/db/models/functions/datetime.py tests/db_functions/datetime/test_extract_trunc.py`

File: tests/db_functions/datetime/test_extract_trunc.py [hunk 1]
```diff
@@ -90,6 +90,26 @@ def create_model(self, start_datetime, end_datetime):
             duration=(end_datetime - start_datetime) if start_datetime and end_datetime else None,
         )
 
+    def assertIsoYearQueryUsesExtraction(self, query):
+        sql = str(query).lower()
+        self.assertIn('extract', sql)
+        self.assertNotIn(' between ', sql)
```
Comment (noa-lucent) [inline tests/db_functions/datetime/test_extract_trunc.py:95]: [major]  assumes every backend emits , but that's only true on PostgreSQL. MySQL formats ISO years via , Oracle uses , and SQLite routes through the  UDF. This assertion will immediately fail on those backends and it doesn't actually prove that we're compiling to the backend-specific ISO-year expression. Please make this helper backend-aware (e.g., branch on  or, better, inspect the compiled where node and assert that its LHS is an  instance) while still checking that  is absent.}

Comment (noa-lucent) [inline tests/db_functions/datetime/test_extract_trunc.py:95]: [major] `assertIsoYearQueryUsesExtraction()` assumes every backend emits `'extract'`, but only PostgreSQL does. MySQL formats ISO years via `TRUNCATE(YEARWEEK(..., 3), -2) / 100`, Oracle uses `TO_CHAR(..., 'IYYY')`, and SQLite routes through the `django_datetime_extract` UDF. This assertion will fail on those backends and it doesn't actually prove we're compiling to the backend-specific ISO-year expression. Please make this helper backend-aware (e.g., branch on `connection.vendor` or, better, inspect the compiled where node and assert that its LHS is an `ExtractIsoYear` instance) while still checking that `' between '` is absent.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Summary:\n- Verified ExtractIsoYear now registers the standard comparison lookups while the calendar year BETWEEN optimization remains unchanged.\n- Confirmed the PR description includes the requested reproduction details and the lone commit is tagged with [skip ci].\n- Noted the new regression tests cover the implicit transform, annotation pairing, inequalities, and calendar-vs-ISO boundaries.\n\nBlocking issue:\n-  hardcodes a PostgreSQL-specific SQL fragment. On MySQL, Oracle, and SQLite the ISO-year expression looks different, so the test will fail and still wouldn\'t prove the transform compiled correctly. Please make the assertion backend-aware (or inspect the compiled nodes directly) while keeping the BETWEEN check.

File: tests/db_functions/datetime/test_extract_trunc.py (hunk 1 above)
Comment (casey-brooks) [reply tests/db_functions/datetime/test_extract_trunc.py:95]: Updated `assertIsoYearQueryUsesExtraction()` to walk the where-clause tree and verify that every iso_year lookup uses an `ExtractIsoYear` LHS while still asserting the SQL never falls back to `BETWEEN`. This keeps the check backend-agnostic and matches the requested guard.

**Review by noa-lucent — APPROVED** [review]
Looks great. The iso-year helper now walks the where-clause tree, so we verify  is used regardless of backend SQL spelling while still guarding against the BETWEEN fallback.
